### PR TITLE
Implement the conflict-free scheduler

### PR DESCRIPTION
**Pull request by yongla**
This is the core implementation of the `CourseScheduler` class which generates conflict-free schedules using a SAT solver.

Given the user's saved courses, the scheduler sets the following constraints to the SAT solver:
 - Allow only 1 section of a lecture, tutorial, lab, and/or seminar for a course
 - Include all classes of a section (e.g. M, W, and F lectures) to be part of the solution
 - Include all types of classes to be part of the solution (a LEC, a TUT, a LAB, etc.)
 - Disallow any conflicting pairs of lectures, tutorials, labs, and seminars from the solution

The call `generateSchedules()` will create the constraints and generate conflict-free schedules. Use the call `getCurrentSchedule()` to retrieve a `List<CourseComponent>` that represent a conflict-free schedule with all the user's courses.

More functionality will be added in the future, but this is the foundation of the scheduler

### CODE REVIEW

File: app/src/main/java/com/example/peterchu/watplanner/scheduler/CourseScheduler.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+     * Adds constraints to the solver
+     */
+    private List<List<BoolVar>> addConstraints(List<CourseComponent> lectures)
+            throws ParseException {
+        List<List<BoolVar>> sectionList = getBoolListsBySection(lectures);
+
+        if (sectionList.size() == 1) {
+            List<BoolVar> sectionA = sectionList.get(0);
+            BoolVar[] a = sectionA.toArray(new BoolVar[sectionA.size()]);
+            solver.post(and(a));
```
Comment (timothytong) [inline app/src/main/java/com/example/peterchu/watplanner/scheduler/CourseScheduler.java:125]: Is it acceptable to have only 1 arg passed into `and`?

Comment (yongla) [reply app/src/main/java/com/example/peterchu/watplanner/scheduler/CourseScheduler.java:125]: `a` is an array of `BoolVar`s that are all `and`ed together

File: app/src/main/java/com/example/peterchu/watplanner/scheduler/CourseScheduler.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+    }
+
+    public List<CourseComponent> getCurrentSchedule() {
+        return currentSchedule;
+    }
+
+    /**
+     * Adds constraints to the solver
+     */
+    private List<List<BoolVar>> addConstraints(List<CourseComponent> lectures)
```
Comment (timothytong) [inline app/src/main/java/com/example/peterchu/watplanner/scheduler/CourseScheduler.java]: Are these going to be just `lectures`? Or a generic `courseComponents`

**Review by PeterChu2 — APPROVED** [review]
DAMN lgtm! just in time for our oral defence

**Comment by Mrez95** [comment]
🚢 🛳 ⛴ ⛵️🚤 🛥 🛶 🚣🏿

File: app/src/main/java/com/example/peterchu/watplanner/scheduler/CourseScheduler.java (hunk 2 above)
Comment (yongla) [reply app/src/main/java/com/example/peterchu/watplanner/scheduler/CourseScheduler.java]: could be tutorials or labs and stuff too. i'll update the name before merging

**Pull request merged**